File: lib/services/ideasStatsService.js
/**
 * @fileoverview Hilfsfunktionen für Statistik-Abfragen (z.B. Wochenquoten für Likes/Dislikes).
 * JSDoc in diesem Projekt ist auf Deutsch verfasst.
 * @module lib/services/ideasStatsService
 */

const db = require('../../config/db.js');

async function getWeeklyRemaining(userId) {
  const weeklyLikeRows = await db.query(
    'SELECT COUNT(*) AS cnt FROM likes WHERE user_id = ? AND created_at >= DATE_SUB(NOW(), INTERVAL 7 DAY)',
    [userId]
  );
  const weeklyLikes = weeklyLikeRows[0]?.cnt;

  const weeklyDislikeRows = await db.query(
    'SELECT COUNT(*) AS cnt FROM dislikes WHERE user_id = ? AND created_at >= DATE_SUB(NOW(), INTERVAL 7 DAY)',
    [userId]
  );
  const weeklyDislikes = weeklyDislikeRows[0]?.cnt;

  return {
    remainingLikes: Math.max(0, 3 - (Number(weeklyLikes) || 0)),
    remainingDislikes: Math.max(0, 3 - (Number(weeklyDislikes) || 0))
  };
}

module.exports = {
  getWeeklyRemaining
};

File: newelixirideaboard/lib/idea_board/services/ideas_stats_service.ex
defmodule IdeaBoard.IdeasStatsService do
  def get_stats(idea_id) do
    {:ok, result} = IdeaBoard.Repo.query(
      "SELECT IFNULL(like_count, 0) AS like_count, IFNULL(dislike_count, 0) AS dislike_count, IFNULL(comment_count, 0) AS comment_count, score FROM ideas WHERE idea_id = ?",
      [idea_id]
    )
    result.rows[0]
  end

  def get_weekly_remaining(user_id) do
    {:ok, result} = IdeaBoard.Repo.query(
      "SELECT COUNT(*) AS count FROM ideas WHERE user_id = ? AND created_at >= DATE_SUB(NOW(), INTERVAL 7 DAY)",
      [user_id]
    )
    result.rows[0] && result.rows[0][0] || 0
  end
end
